src/Sema/LowerZon.zig

```diff
@@ -0,0 +1,902 @@
+        },
+    }
+}
+
+fn lazySrcLoc(self: *LowerZon, loc: LazySrcLoc.Offset) !LazySrcLoc {
+    if (self.base_node_inst == .none) {
+        self.base_node_inst = (try self.sema.pt.zcu.intern_pool.trackZir(
+            self.sema.pt.zcu.gpa,
+            .main,
+            .{ .file = self.file_index, .inst = .main_struct_inst },
+        )).toOptional();
+    }
+    return .{
+        .base_node_inst = self.base_node_inst.unwrap().?,
+        .offset = loc,
+    };
+}
+
+fn failUnsupportedResultType(
+    self: *LowerZon,
+    msg: []const u8,
+    ty: Type,
+) error{ AnalysisFail, OutOfMemory } {
+    @branchHint(.cold);
+    const err_msg = try Zcu.ErrorMsg.create(self.sema.pt.zcu.gpa, self.import_loc, "{s}", .{msg});
+    try self.sema.pt.zcu.errNote(
+        self.import_loc,
+        err_msg,
+        "result type contains type '{}'",
+        .{ty.fmt(self.sema.pt)},
+    );
+    return self.sema.failWithOwnedErrorMsg(self.block, err_msg);
+}
+
+fn fail(
+    self: *LowerZon,
+    node: Zoir.Node.Index,
+    comptime format: []const u8,
+    args: anytype,
+) error{ AnalysisFail, OutOfMemory } {
+    @branchHint(.cold);
+    const src_loc = try self.lazySrcLoc(.{ .node_abs = node.getAstNode(self.file.zoir.?) });
+    const err_msg = try Zcu.ErrorMsg.create(self.sema.pt.zcu.gpa, src_loc, format, args);
+    try self.sema.pt.zcu.errNote(self.import_loc, err_msg, "imported here", .{});
+    return self.sema.failWithOwnedErrorMsg(self.block, err_msg);
+}
+
+fn lowerExpr(self: *LowerZon, node: Zoir.Node.Index, res_ty: Type) CompileError!InternPool.Index {
+    const pt = self.sema.pt;
+    return self.lowerExprInner(node, res_ty) catch |err| switch (err) {
+        error.WrongType => {
+            // Flatten to a type directly compatible with ZON before reporting the error
+            var display_ty = res_ty;
+            var opt = false;
+
+            while (true) {
+                switch (display_ty.zigTypeTag(pt.zcu)) {
+                    .optional => {
+                        opt = true;
+                        display_ty = display_ty.childType(pt.zcu);
+                    },
+                    .pointer => {
+                        const ptr_info = res_ty.ptrInfo(pt.zcu);
+                        if (ptr_info.flags.size == .one) {
+                            display_ty = .fromInterned(ptr_info.child);
+                        } else {
+                            break;
+                        }
+                    },
+                    else => break,
+                }
+            }
+
+            if (opt) {
+                display_ty = .fromInterned(try pt.intern(.{
+                    .opt_type = display_ty.toIntern(),
+                }));
+            }
+
+            // Report the error
+            return self.fail(
+                node,
+                "expected type '{}'",
+                .{display_ty.fmt(self.sema.pt)},
+            );
+        },
+        else => |e| return e,
```

Comment:
I don't actually think it makes sense to do this in the comptime ZON parser. The use cases are different here:
* Errors in the runtime parser are likely to be seen by end users, so should correlate to the ZON they should write
* Errors in the comptime parser are going to be seen by programmers, so should correlate to the actual type being lowered (also for consistency with other error messages)

Reply:
Good point--I agree with this.